internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,340 @@
+	mu                 sync.Mutex
+	opThrottle         *throttle.Throttle
+	parallelChunkLimit int
+	sizeLimit          int64
+}
+
+type Info struct {
+	ID        string // upload operation identifier. Ephemeral, just used for the upload process
+	DocID     string // document ID of the uploaded file and chunks
+	Source    string // which integration is performing the upload
+	ChunkSize int64
+	Total     int64
+	Count     int
+	HashSum   string
+	Hasher    hash.Hash
+}
+
+type ChunkInfo struct {
+	ID            int
+	FirstReceived bool
+	Final         bool
+	Upload        Info
+	Token         *throttle.Token
+}
+
+func New(sizeLimit int64, opLimit int, chunkLimit int) *Uploader {
+	return &Uploader{
+		parallelChunkLimit: chunkLimit,
+		sizeLimit:          sizeLimit,
+		opThrottle:         throttle.NewThrottle(opLimit),
+		current:            make(map[string]upload, opLimit),
+	}
+}
+
+// Start an upload operation, as long as the max concurrent has not been reached
+// returns the upload ID
+func (u *Uploader) Begin(size int64, docID string, source string, hashsum string, hasher hash.Hash) (Info, error) {
+	if size <= 0 {
+		return Info{}, errors.New("invalid file size")
+	}
+	if size > u.sizeLimit {
+		return Info{}, ErrFileSizeTooLarge
+	}
+
+	uid, err := uuid.NewV4()
+	if err != nil {
+		return Info{}, fmt.Errorf("unable to generate upload operation ID: %w", err)
+	}
+	id := uid.String()
+
+	token := u.opThrottle.Acquire(id, 300*time.Hour)
+	if token == nil {
+		return Info{}, ErrMaxConcurrentUploads
+	}
+
+	total := time.NewTimer(uploadRequestTimeout)
+	chunkT := time.NewTimer(chunkProgressTimeout)
+	chunkRecv := make(chan struct{})
+	complete := make(chan struct{})
+	// total timer could also be achieved with context deadline and cancelling
+
+	go func() {
+		for {
+			select {
+			case <-total.C: // entire upload operation timed out
+				log.Trace().Str("uploadID", id).Msg("upload operation timed out")
+				// stop and drain chunk timer
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				u.cancel(id)
+				return
+			case <-chunkT.C: // no chunk progress within chunk timer, expire operation
+				log.Trace().Str("uploadID", id).Msg("upload operation chunk activity timed out")
+				// stop and drain total timer
+				if !total.Stop() {
+					<-total.C
+				}
+				u.cancel(id)
+				return
+			case <-chunkRecv: // chunk activity, update chunk timer
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				chunkT.Reset(chunkProgressTimeout)
+			case <-complete: // upload operation complete, clean up
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				if !total.Stop() {
+					<-total.C
+				}
+				u.finalize(id)
+				return
+			}
+		}
+	}()
+	info := Info{
+		ID:        id,
+		DocID:     docID,
+		ChunkSize: MaxChunkSize,
+		Source:    source,
+		Total:     size,
+		Hasher:    hasher,
+		HashSum:   hashsum,
+	}
+	cnt := info.Total / info.ChunkSize
+	if info.Total%info.ChunkSize > 0 {
+		cnt += 1
+	}
+	info.Count = int(cnt)
+	u.current[id] = upload{
```

Comment:
needs mutex access

Reply:
the initial write should be fine.  `id` will be some freshly generated random value so there is near zero chance its overwriting anything existing. Just adding a new value. And no client has been told the `id` yet from any responses, so nothing should be concurrently checking its value

Reply:
Just to confirm, concurrent access to this map is impossible? Concurrent access to a map will panic so I would recommend being defensive with a mutex if there is any possibility of concurrent access here.

Reply:
Logically impossible, but I'm happy to wrap this in a lock to be safe